### PR TITLE
Do not set Content-length header when trailing headers exists

### LINKED ISSUE #1554: HttpResponse's trailing headers are not sent.
Originally reported by @southernkasaist

```java
// Trailer is sent.
ctx.addAdditionalResponseTrailer(HttpHeaderNames.WWW_AUTHENTICATE,
                                 "Basic realm=\"Authentication required for this page\"");
return HttpResponse.of(HttpStatus.UNAUTHORIZED,
                       MediaType.HTML_UTF_8,
                       HttpData.ofUtf8("401 Forbidden"));

// Trailer is not set.
return HttpResponse.of(HttpStatus.UNAUTHORIZED,
                       MediaType.HTML_UTF_8,
                       HttpData.ofUtf8("401 Forbidden"),
                       HttpHeaders.of(HttpHeaderNames.WWW_AUTHENTICATE,
                                      "Basic realm=\"Authentication required for this page\""));
```

In the second case, after writing `HttpData.ofUtf8("401 Forbidden")`, connection was closed and the trailing headers of `WWW-Authenticate` was subscribed but not written to response.

### PR DESCRIPTION
**Pull request by minwoox**
Motivations:
Some of the client implementations such as `curl` ignores trailing headers if the `Content-length` header is present.
We should not set `Content-length` header when trailing headers exists so that those clients can receive the trailing headers.

Modification:
- Do not set `Content-length` header when trailing headers exists
- Make a backlist which must not be included in trailing headers
  - https://tools.ietf.org/html/rfc7230#section-4.1.2
- Allow `Trailer` headers in response headers which was not before
- Set `Content-length` to `0` when the `HttpResponse` contains only a status so that the response is not transfer encoding

Result:
- Closes #1554

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/HttpResponse.java [hunk 1]
```diff
@@ -255,20 +259,32 @@ static HttpResponse of(HttpHeaders headers, HttpData content, HttpHeaders traili
...
-        if (isContentAlwaysEmptyWithValidation(status, content, trailingHeaders)) {
+        if (isContentAlwaysEmptyWithValidation(status, content, trailingHeaders) ||
+            (content.isEmpty() && trailingHeaders.isEmpty())) {
             ReferenceCountUtil.safeRelease(content);
-            return new OneElementFixedHttpResponse(headers);
-        } else if (!content.isEmpty()) {
-            if (trailingHeaders.isEmpty()) {
-                return new TwoElementFixedHttpResponse(headers, content);
+
+            final HttpHeaders mutableHeaders = headers.toMutable();
```
Comment (hyangtack) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponse.java]: nit: How about calling `toMutable()` inside the below `if` statement?

**Comment by minwoox** [comment]
All fixed. Thank you!

File: core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java [hunk 2]
```diff
@@ -665,6 +723,57 @@ public static void toNettyHttp1(
...
+                                                       HttpHeaders trailingHeaders) {
+        requireNonNull(headers, "headers");
+        requireNonNull(content, "content");
+        requireNonNull(trailingHeaders, "trailingHeaders");
+
+        final HttpStatus status = headers.status();
+        assert status != null;
+
+        final HttpHeaders mutable = headers.toMutable();
+        if (isContentAlwaysEmptyWithValidation(status, content, trailingHeaders)) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java]: I had to put this first to validate the status. >_<

File: core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java [hunk 3]
```diff
@@ -723,6 +723,54 @@ public static void toNettyHttp1(
         }
     }
 
+    /**
+     * Returns a {@link HttpHeaders} which {@link HttpHeaderNames#CONTENT_LENGTH} is added or removed
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java]: - a -> an
- which -> whose

File: core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java [hunk 4]
```diff
@@ -583,12 +626,24 @@ private static void setHttp2Scheme(io.netty.handler.codec.http.HttpHeaders in, U
...
      */
-    public static Http2Headers toNettyHttp2(HttpHeaders in) {
+    public static Http2Headers toNettyHttp2(HttpHeaders in, boolean server) {
         final Http2Headers out = new DefaultHttp2Headers(false, in.size());
-        out.set(in);
-        out.remove(HttpHeaderNames.CONNECTION);
-        out.remove(HttpHeaderNames.TRANSFER_ENCODING);
-        out.remove(HttpHeaderNames.TRAILER);
+
+        // Trailering headers if it does not have :status.
```
Comment (hyangtack) [inline core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java]: nit: Trailering -> Trailing

**Comment by trustin** [comment]
Thanks, @minwoox !

**Pull request merged**